pkg/controller/node/nodecontroller.go

```diff
@@ -503,19 +526,49 @@ func (nc *NodeController) monitorNodeStatus() error {
 			// Check eviction timeout against decisionTimestamp
 			if observedReadyCondition.Status == api.ConditionFalse &&
 				decisionTimestamp.After(nc.nodeStatusMap[node.Name].readyTransitionTimestamp.Add(nc.podEvictionTimeout)) {
-				if nc.evictPods(node) {
-					glog.V(2).Infof("Evicting pods on node %s: %v is later than %v + %v", node.Name, decisionTimestamp, nc.nodeStatusMap[node.Name].readyTransitionTimestamp, nc.podEvictionTimeout)
+				nodeNotReadyTaint := api.Taint{
+					Key:       unversioned.TaintNodeNotReady,
+					Effect:    api.TaintEffectNoExecute,
+					AddedTime: nc.now(),
+				}
+				added, err := tryAddTaintToNode(nc.kubeClient, node.Name, nodeNotReadyTaint)
+				if err != nil {
+					glog.Errorf("Failed to try add taint %s to node %s: %v", nodeNotReadyTaint.ToString(), node.Name, err)
+				}
+				if added {
+					glog.V(4).Infof("Added nodeNotReady taint to node %s: %v is later than %v + %v", node.Name, decisionTimestamp, nc.nodeStatusMap[node.Name].readyTransitionTimestamp, nc.podEvictionTimeout)
 				}
 			}
 			if observedReadyCondition.Status == api.ConditionUnknown &&
 				decisionTimestamp.After(nc.nodeStatusMap[node.Name].probeTimestamp.Add(nc.podEvictionTimeout)) {
-				if nc.evictPods(node) {
-					glog.V(2).Infof("Evicting pods on node %s: %v is later than %v + %v", node.Name, decisionTimestamp, nc.nodeStatusMap[node.Name].readyTransitionTimestamp, nc.podEvictionTimeout-gracePeriod)
+				nodeUnreachableTaint := api.Taint{
+					Key:       unversioned.TaintNodeUnreachable,
+					Effect:    api.TaintEffectNoExecute,
+					AddedTime: nc.now(),
+				}
+				added, err := tryAddTaintToNode(nc.kubeClient, node.Name, nodeUnreachableTaint)
```

Comment:
NodeReady can have three values: True, False and Unknown. True means ready, False and Unknown means that something's wrong. False is generally set by Kubelet to mean that something's wrong (e.g. Docker is down). Unknown means that the Node is Unreachable. Those are different things and I don't understand why we shouldn't have both taints.